src/build_helper/src/git.rs

```diff
@@ -173,7 +176,177 @@ pub fn get_closest_merge_commit(
     Ok(output_result(&mut git)?.trim().to_owned())
 }
 
+/// Represents the result of checking whether a set of paths
+/// have been modified locally or not.
+#[derive(PartialEq, Debug)]
+pub enum PathFreshness {
+    /// Artifacts should be downloaded from this upstream commit,
+    /// there are no local modifications.
+    LastModifiedUpstream { upstream: String },
+    /// There are local modifications to a certain set of paths.
+    /// "Local" essentially means "not-upstream" here.
+    /// `upstream` is the latest upstream merge commit that made modifications to the
+    /// set of paths.
+    HasLocalModifications { upstream: String },
+}
+
+/// This function figures out if a set of paths was last modified upstream or
+/// if there are some local modifications made to them.
+///
+/// It can be used to figure out if we should download artifacts from CI or rather
+/// build them locally.
+///
+/// `target_paths` should be a non-empty slice of paths (relative to `git_dir` or the
+/// current working directory) whose modifications would invalidate the artifact.
+/// Each path can also be a negative match, i.e. `:!foo`. This matches changes outside
+/// the `foo` directory.
```

Comment:
Question (wording): wait so this is saying that given a path filter set `{":!foo"}` only changes to files matching outside of `foo` directory can be eligible for downloading upstream artifact?

- Can `foo` be a file (non-directory)?
- Can `foo` be a symlink?

Reply:
> wait so this is saying that given a path filter set {":!foo"} only changes to files matching outside of foo directory can be eligible for downloading upstream artifact?

Yes, I think so (it's functionality of git, not bootstrap). It's described [here](https://git-scm.com/docs/gitglossary#Documentation/gitglossary.txt-aiddefpathspecapathspec).

Reply:
Right. Maybe just dump a git doc link here (unless there's one already)?

Reply:
Good idea, will add it.